src/C50A0.c

```diff
@@ -158,9 +192,55 @@ INCLUDE_ASM(s32, "C50A0", func_80132D94);
 
 INCLUDE_ASM(s32, "C50A0", render_item_entities);
 
-INCLUDE_ASM(s32, "C50A0", remove_item_entity_by_reference);
+void remove_item_entity_by_reference(ItemEntity* entity) {
+    s32 index;
+
+    for (index = 0; index < 0x100; index++) {
+        if (D_801565A0[index] == entity) {
+            break;
+        }
+    }
+
+    if (index < 0x100) {
+        if (entity->physicsData != NULL) {
+            heap_free(entity->physicsData);
+        }
+
+        switch ((s8)entity->type) {
+            case 0:
+            case 3:
+            case 12:
+            case 16:
+            case 20:
+            case 28:
+                delete_shadow(entity->shadowIndex);
+                break;
+        }
+
+        heap_free(D_801565A0[index]);
+        D_801565A4 = 0;
+        D_801565A0[index] = NULL;
+    }
+}
+
+void remove_item_entity_by_index(s32 index) {
+    ItemEntity* itemEntity = D_801565A0[index];
 
-INCLUDE_ASM(s32, "C50A0", remove_item_entity_by_index);
+    switch ((s8)itemEntity->type) {
```

Comment:
Should the type of `itemEntity->type` be changed to s8? How wide is it now?

Reply:
Changed the type to s8